File: 3.longest-substring-without-repeating-characters.cs
/*
 * @lc app=leetcode id=3 lang=csharp
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
public class Solution
{
    public int LengthOfLongestSubstring(string s)
    {
        int maxLen = 0;
        int left = 0;
        int right = 0;
        HashSet<char> uniqueChars = new HashSet<char>();

        while (right < s.Length)
        {
            if (!uniqueChars.Contains(s[right]))
            {
                uniqueChars.Add(s[right]);
                maxLen = Math.Max(maxLen, right - left + 1);
                right++;
            }
            else
            {
                uniqueChars.Remove(s[left]);
                left++;
            }
        }
        return maxLen;

    }
}
// @lc code=end

File: 3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#

# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        right = 0
        left = 0
        maxLen = 0
        uniqueChars = set()

        while right < len(s):
            if s[right] in uniqueChars:
                uniqueChars.remove(s[left])
                left = left + 1
            else:
                uniqueChars.add(s[right])
                maxLen = max(maxLen, right - left + 1)
                right = right + 1
        
        return maxLen

# @lc code=end

